core/cap-0046-09.md

```diff
@@ -86,18 +86,21 @@ is only being partially updated. This is done to simplify the upgrade format
 (otherwise we'd need to introduce some key-value mechanism of specifying
 upgraded fields).
 
-Multiple configuration setting entries may be upgraded at once using
+Multiple configuration setting entries may be upgraded at once using a
 `ConfigUpgradeSet` struct that contains all the updated `ConfigSettingEntry`
-entries. Similar to the transaction sets, the network needs to agree on hash of
-the `ConfigUpgradeSet`, hence only hash is added to `SCPValue`. While upgrades
-are rare events, the size of the setting entry is not bounded and the number of
-`SCPValues` in the network is large enough (thousands of messages per node as
-of writing this) to have a risk of significantly reducing the network bandwidth
-during the upgrade.
-
-The actual upgrade set contents are pulled from the peers in a similar fashion
-to the transaction sets (if needed at all; ideally all the validators would
-have exactly the same upgrade set).
+entries. The `LedgerUpgrade` in `StellarValue` that validators will vote on will
+contain only an `ConfigUpgradeSetKey`, which containts a contractID, and the
+SHA-256 hash of the `ConfigUpgradeSet`. The actual `ConfigUpgradeSet` will need
+to exist on ledger in a `ContractDataEntry`, and the validators will look it up
+using the `ConfigUpgradeSetKey`. This means that a WASM contract will need to be
+used to write the proposed upgrade. The specs of the entry written is as follows -
+- It should be a `TEMPORARY` entry, so make sure it's bumped to live long enough for the upgrade.
+- The `SCVal` `key` should be of type `SCV_BYTES`, where the value is the SHA-256 hash of the `ConfigUpgradeSet`.
+- The `SCVal` `val` should also be of type `SCV_BYTES`, where the value is the
+  serialized XDR of the `ConfigUpgradeSet`.
+  
+Once the proposed upgrade is setup, you'll need to share the
+`ConfigUpgradeSetKey` with the other validators so they can vote for it.
```

Comment:
I think we should document somewhere that core validates the nominated entry by verifying that the value hash equals to the key hash.

Reply:
Done